dhcp-controller/dhcpserver/pkg/server/server.go

```diff
@@ -0,0 +1,400 @@
+		dirInit(leasePath)
+		StartWatcher(lf, leasePath)
+	}
+}
+
+func serverStart(dnsmasqBinary string, args []string) *exec.Cmd {
+	cmd := &exec.Cmd{
+		Path: dnsmasqBinary,
+		Args: args,
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		panic(err)
+	}
+	if err := cmd.Start(); err != nil {
+		panic(err)
+	}
+	buf := bufio.NewReader(stderr)
+	go func() {
+		for {
+			line, _, err := buf.ReadLine()
+			if err != nil {
+				break
+			}
+			dnsmasqLog.Info(string(line))
+		}
+	}()
+	return cmd
+}
+
+// Retrieve leases from backup. Will be retrieved from etcd later on
+func retrieveBackup(leasePath string) error {
+	serverLog.Info("Retrieving Backup")
+
+	ipAllocations, err := k8sClient.ListIPAllocations(context.TODO())
+	if err != nil {
+		serverLog.Error(err, "failed to fetch IPallocations")
+	}
+
+	destination, err := os.Create(leasePath)
+	if err != nil {
+		return err
+	}
+	defer destination.Close()
+
+	for _, ipallocation := range ipAllocations {
+		// Restore IPs that originated from this pod
+
+		//TODO do we need to store epochTime as well
+		for idx, mask := range Netmask {
+			length, _ := net.IPMask(net.ParseIP(mask).To4()).Size()
+			ipv4Addr := net.ParseIP(ipallocation.Name)
+			ipv4Mask := net.CIDRMask(length, 32)
+
+			if HostNetwork[idx].String() == ipv4Addr.Mask(ipv4Mask).String() {
+				if err != nil {
+					serverLog.Error(err, "failed to get lease time")
+				}
+				// This loop only runs once
+				for ip, obj := range ipallocation.Spec.Allocations {
+					tmpline := fmt.Sprintf(ipallocation.Spec.EpochExpiry + " " + obj.MacId + " " + ip + " " + obj.VmiRef + " *\n")
+					destination.WriteString(tmpline)
+				}
+				serverLog.Info("Restored IPAllocation " + ipallocation.Name)
+			}
+		}
+
+	}
+
+	return err
+}
+
+// Check if the lease file exists
+func dirInit(leasePath string) {
+	_, err := os.Stat(leasePath)
+	if os.IsNotExist(err) {
+		// If not, retrieve from backup and create leasefile
+		retrieveBackup(leasePath)
+	}
+}
+
+// Checks if lease found in records exist in leasefile.
+// This is for the scenario when leases expire and dnsmasq deletes
+// the lease from the leasefile
+func leaseExist(ip string, records [][]string) (result bool) {
+	result = false
+	for _, lease := range records {
+		if lease[2] == ip {
+			result = true
+			break
+		}
+	}
+	return result
+}
+
+// Checks if existing entry in leasefile has been updated
+func checkRecord(lease LeaseFile, record []string) bool {
+	serverLog.Info("Checking lease entry....")
+	var isupdated = false
+	if record[0] != lease.EpochTimestamp ||
+		record[1] != lease.MacAddress ||
+		record[2] != lease.IPAddress ||
+		record[3] != lease.Hostname ||
+		record[4] != lease.ClientID {
+		isupdated = true
+	}
+	return isupdated
+}
+
+// updates record with new data from leasefile
+func updateRecord(lf map[string]LeaseFile, record []string, isupdate bool) {
+	serverLog.Info("Updating Lease")
+	lf[record[2]] = LeaseFile{record[0], record[1], record[2], record[3], record[4]}
```

Comment:
This is a little cryptic, it isn't clear what these fields represent. Can we change it to a typed struct or a map with readable keys, rather than an array?

For example, record["mac"], record["IP"] or record.Mac, record.Ip

Reply:
Done. Changed it to a typed struct.